source/crud/crud.rst

```diff
@@ -1927,8 +1928,72 @@ The ``update`` family of operations require that the update document parameter M
 The ``replace`` family of operations require that the replacement document parameter MUST NOT begin with an atomic modifier. In practice, this means that introspection needs to happen on that document to enforce this. However, it is enough to only check the first element in the document. If it does not begin with a ``$`` sign but an element later on does, the server will throw an error.
 
 
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Aggregation Pipelines with Write Stages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This section discusses special considerations for aggregation pipelines that
+contain write stages (e.g. ``$out``, ``$merge``).
+
+
+Returning a cursor on the output collection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As noted in the documentation for the ``aggregate`` helper earlier in this
+document, ``$out`` and ``$merge`` are special pipeline stages that cause no
+results to be returned from the server. As such, drivers MAY setup a cursor to
+be executed upon iteration against the output collection and return that instead
+of an iterable that would otherwise have no results.
+
+Drivers that do so for ``$merge`` MAY remind users that such a cursor may return
+more documents than were written by the aggregation (e.g. documents that existed
+in the collection prior to ``$merge`` being executed).
+
+
+Read preferences and server selection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Historically, only primaries could execute an aggregation pipeline with ``$out``
+or ``$merge``. As of ``featureCompatibilityVersion`` 4.4, secondaries can now
+execute such pipelines; however, since drivers do not track
+``featureCompatibilityVersion``, executing an aggregation pipeline with ``$out``
+or ``$merge`` on a secondary will require a 5.0+ server (i.e. wire version >=
+13).
+
+Drivers MUST guarantee the following:
+
+- If an explicit (i.e. per-operation) read preference is specified for an
+  aggregation with a write stage, drivers MUST attempt to use it. If that would
+  result in a pre-5.0, secondary server being selected, drivers MUST instead
+  select a server using a primary read preference.
```

Comment:
Not sure exactly how server selection works in other drivers, but we use filtering functions in the Go driver: lists of candidate servers are narrowed down based on certain criteria (latency, read preference, etc.). It's not exactly clear _which_ server will be selected; we just remove servers from the candidates list if they do not fit the criteria. So, I interpreted this (and similar statements below) as: `remove all pre-5.0 secondary servers from the candidates list. If no candidates remain, fall back to the primary.`

Still a couple questions here, though:
1. What is the expected behavior for an aggregate with an output stage with `secondary` read preference against a single-node replica set (i.e. only one node and it's a primary)? Is this a server selection error?
2. If the user specifies `secondary` read preference, but we fall back to the primary because there are no usable secondaries, what is the `$readPreference` sent to the server in the actual command? `secondary` or `primary`?

Reply:
> [in this case] I would always expect a `secondary` RP to result in a server selection failure.

Same here, and I agree that "fixing" server selection in this case will only further disrupt our process of server selection.

> `$readPreference` in the OP_MSG command should always correspond to the _effective_ read preference that was used.

This sounds good to me, and I do think the spec should clarify that "falling back to primary" changes both the effective RP and the `$readPreference` in the command to be `primary`. Is it possible to test that case with a spec test?

Reply:
> I do think the spec should clarify that "falling back to primary" changes both the effective RP and the `$readPreference` in the command to be primary. Is it possible to test that case with a spec test?

I can amend this language to describe the concept of an _effective_ read preference. As far as testing, I think the current tests that expect `$readPreference` to be absent is already covering the "fall back to primary" case. A `primary` read preference never gets forwarded to the server, so its absence in the command document is how we assert that the command is destined for a primary (either directly in a replica set topology or via mongos/LB).